### ISSUE TITLE
npm install doesn't update to minimum required version in devDependencies

### ISSUE THREAD

**Issue by mkeremguc** [npm/npm#8008]
say in package.json:
```
"devDependencies": {
    "my-package": "~1.0.0",
}
```
so when the version is bumped to `~1.0.1` in package.son, it should update my-package to 1.0.1 when `npm install` is run. it skips the minimum version check and skips install and keeps the version as 1.0.0 there..

using npm v2.8.3. it was working till npm version 2.6 or so

**Comment by smikes** [comment]
Thanks for the bug report.  Can you run `npm install -ddd` and upload the resulting npm-debug.log to https://gist.github.com and link it here?  That will help us narrow down the cause.

**Comment by rocketraman** [comment]
:+1: Same problem here. Seems to be working fine with npm 2.7.5, and seems to be broken in npm 2.7.6  and npm 2.8.4. I didn't test versions between 2.7.6 and 2.8.4.

With version 2.8.4 the initial state is as follows (note that react and react-router are in `dependencies` and react-tools is in `devDependencies`:

```
$ npm ls react react-router react-tools
├── react@0.12.2 
├── react-router@0.12.2 
└── react-tools@0.12.2
```

and then updating package.json to react 0.13.2 for all three of the above projects and running `npm install -ddd`. See https://gist.github.com/rocketraman/ca381cb12c18e4ad99b5.

Output of `npm ls` after:

```
$ npm ls react react-router react-tools
├── react@0.13.2 
├── react-router@0.13.2 
└── react-tools@0.12.2 
```

**Comment by smikes** [comment]
Confirmed, thanks for the specific example.  A test repo is here: git@github.com:smikes/npm-8008-test.git

Currently working on narrowing down the cause.

**Comment by smikes** [comment]
For some reason, the "already installed" logic is getting tripped, and `npm` is skipping the dev dependency claiming:

```
npm verb already installed skipping react-tools@^0.13.2 /home/raman/source/myproject/web
```

**Comment by smikes** [comment]
Looks like in npm@2.7.5 `targetResolver` is getting consolidated deps (includes devdeps) but in npm@2.8.4, `targetResolver` has `undefined` as requested dep for react-tools...

**Comment by smikes** [comment]
I have a candidate fix in https://github.com/npm/npm/pull/8018

**Comment by othiym23** [comment]
@smikes's fix landed in 9ad210042242e51d52b2a8b633d8e59248f5faa4. Thanks for reporting this!
